### ISSUE TITLE
http redirect speed issues.

### ISSUE THREAD

**Issue by Kr3m** [ec-/Quake3e#206]
Not really sure what is going on, but no matter what I do with my settings, my redirect url is capped at less than 2 MB/s.  On my UT2004 server, using the same redirect, I get closer to 30 MB/s.  What am I missing?  I apologize if this isn't a server related issue with the code and is something in my settings.  I'm just at a loss and can't find anything so I assume it's an issue with the server binary itself.

wget output:

Saving to: 'dt_tourneymappak.pk3'

100%[===>] 96,161,821  26.6MB/s   in 3.9s
2023-01-24 17:48:56 (23.5 MB/s) - 'dt_tourneymappak.pk3' saved [96161821/96161821]

**Comment by ensiform** [comment]
try with curl instead of wget for comparison, as this uses curl.

http redirect has nothing to do with game server binary, its entirely client side engine. server merely provides a url in a cvar.

**Comment by BidyBiddle** [comment]
He might be talking about this: https://github.com/ioquake/ioq3/pull/313

**Comment by Kr3m** [comment]
> try with curl instead of wget for comparison, as this uses curl.
> 
> http redirect has nothing to do with game server binary, its entirely client side engine. server merely provides a url in a cvar.
> 
> Kinda feel like you could be asking these as questions on discord before posting as an issue here as soon as something comes up tbh.

Is there a Discord channel you'd recommend to ask on?  I have another issue that is related to Windows servers that I haven't addressed yet for the very reasons you just suggested.  Linux servers work great.  Not so much on Windows.  One of our guys had to go back to ioq3 to get it working, but I figured I'd hold off on posting that one.

**Comment by Kr3m** [comment]
curl output:

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 91.7M  100 91.7M    0     0  22.9M      0  0:00:03  0:00:03 --:--:-- 22.9M

**Comment by Kr3m** [comment]
> He might be talking about this: [ioquake/ioq3#313](https://github.com/ioquake/ioq3/pull/313)

That definitely seems to be the issue.  I wonder why the branch was deleted and why no one has thought about it since.

**Comment by ensiform** [comment]
> > He might be talking about this: [ioquake/ioq3#313](https://github.com/ioquake/ioq3/pull/313)
> 
> That definitely seems to be the issue. I wonder why the branch was deleted and why no one has thought about it since.

Branches are deleted when merged. Note that cURL documentation suggests this change is only a suggestion to the curl library itself for which it may ignore at any time. But that is a completely separate project, ioq3.

> > try with curl instead of wget for comparison, as this uses curl.
> > http redirect has nothing to do with game server binary, its entirely client side engine. server merely provides a url in a cvar.
> > Kinda feel like you could be asking these as questions on discord before posting as an issue here as soon as something comes up tbh.
> 
> Is there a Discord channel you'd recommend to ask on? Honestly, I felt this question was outside of the scope of anything anyone could answer on Discord anyway. Also, I haven't brought these up as soon as the issues come up. I've been experiencing these problems for a very long time, and figured it was time to address them (outside of the q3plus related issue). In fact I haven't posted an issue in a very long time until recently, because I was so annoyed with the response I got on fs_homepath support being removed forcing everyone to set their fs_homepath or muddy up their Quake 3 install which would just leave files behind in the event of an uninstall.
> 
> I have another issue that is related to Windows servers that I haven't addressed yet for the very reasons you just suggested. Linux servers work great. Not so much on Windows. One of our guys had to go back to ioq3 to get it working, but I figured I'd hold off on posting that one.
> 
> curl output:
> 
> % Total % Received % Xferd Average Speed Time Time Time Current Dload Upload Total Spent Left Speed 100 91.7M 100 91.7M 0 0 22.9M 0 0:00:03 0:00:03 --:--:-- 22.9M
> 
> So even that averages 22.9 MB/s. Definitely an issue with the client itself.

Discord server is mentioned in the readme.

fs_homepath support was never "removed" but rather never "added" or "backported from ioq3", as previously mentioned in other discussion.  The default value of being inside the appdata folder is an ioq3 feature only for windows. Linux and macOS have always used separate paths in all q3a. But you can set it to whatever you wish, in your shortcut or a batch file that you run from shortcut. Since q3plus has own client, `USE_PROFILES`can be defined and used to make it default in that client too. See: https://github.com/ec-/Quake3e/blob/master/code/win32/win_shared.c#L104

`quake3e.x64.exe +set fs_homepath "%APPDATA%\Quake3"`

**Comment by Kr3m** [comment]
> fs_homepath support was never "removed" but rather never "added" or "backported from ioq3", as previously mentioned in other discussion. The default value of being inside the appdata folder is an ioq3 feature only for windows. Linux and macOS have always used separate paths in all q3a. But you can set it to whatever you wish, in your shortcut or a batch file that you run from shortcut. Since q3plus has own client, `USE_PROFILES`can be defined and used to make it default in that client too. See: https://github.com/ec-/Quake3e/blob/master/code/win32/win_shared.c#L104
> 
> `quake3e.x64.exe +set fs_homepath "%APPDATA%\Quake3"`

I'm aware of the differences between Windows and Linux clients.  I just don't understand why the separation support was removed forcing users to set a homepath explicitly.  That was my primary argument.  It's like you took a superior organizational system and dumped it so it would be just as shitty as every other Windows app at keeping anything separate.

**Comment by Kr3m** [comment]
It is nice to know that Linux's default fs_homepath settings were not also removed.  I assumed if you canned the Windows support, you probably did the same thing for Linux so I've been setting that explicitly too.  Just removed it and it works as expected so at least there's that.

**Comment by The-Gig** [comment]
Just dropping in a link to the homepath "issue": https://github.com/ec-/Quake3e/issues/167, for the sake of completeness.

Back to the topic, while of course, _if something can be done better, that's a good thing_, but 2 MB/s is already a "good enough" value to me, who played Q3A back then when I had a 56k modem so my maximum _theoretical_ download speed was 7 KB/s (my actual speed was closer to 3,5 KB/s, I think). LOL.

**Comment by Kr3m** [comment]
> Just dropping in a link to the homepath "issue": https://github.com/ec-/Quake3e/issues/167, for the sake of completeness.
> 
> Back to the topic, a trivia... while of course, _if something can be done better, that's a good thing_, but 2 MB/s is already a "good enough" value to me, who played Q3A back then when I had a 56k modem so my maximum _theoretical_ download speed was 7 KB/s (my actual speed was closer to 3,5 KB/s, I think). LOL.

Man screw that lol.  I already had cable internet in 1999, and didn't miss waiting on Quake 2 and Quake 1 download speeds.  I'm about to update the other issue.  I found the commit where the +scores bug in q3plus started happening.

**Comment by Kr3m** [comment]
Increasing the max buffer size doesn't seem to have helped.  I'd mentioned in Discord that I'd also tried this and nothing changed.  Tried in both Linux and Windows with recent curl versions and it's still stuck at less than cable internet speeds in 1998.  Seems to be some barrier preventing it from reaching even 3 MB/s.  I think we've just stumbled upon a Quake 3 Kobayashi Maru that isn't solvable.  Thanks for trying though.

![20230130212245_1](https://user-images.githubusercontent.com/290740/215646205-d36e8a67-c8fc-40a9-942a-b70019e93930.jpg)
